### PR TITLE
add several `node:zlib` classes

### PR DESCRIPTION
**Pull request by anonrig**
Lacks tests but before adding more, I think this is getting ready to have an additional pair of eyes.

Remaining todo:

- [ ] Add tests
- [x] Investigate why eslint is causing so much trouble (hence eslint-ignore)

### CODE REVIEW

File: src/node/.eslintrc.json [hunk 1]
```diff
@@ -10,32 +10,32 @@
   ],
   "overrides": [],
   "parserOptions": {
-      "ecmaVersion": "latest",
-      "sourceType": "module",
-      "project": "src/node/tsconfig.json"
-    },
+    "ecmaVersion": "latest",
+    "sourceType": "module",
+    "project": "./tsconfig.json"
```
Comment (npaun) [inline src/node/.eslintrc.json:15]: Should be moved to another PR

File: src/node/tsconfig.json [hunk 2]
```diff
@@ -21,9 +21,7 @@
     "moduleResolution": "node",
     "declaration": true,
     "types": [
-      // todo: consume generated workerd types
-      // "@cloudflare/workers-types"
-      "@types/node",
+      "@types/node"
```
Comment (npaun) [inline src/node/tsconfig.json:24]: Also not related

File: src/node/internal/validators.ts [hunk 3]
```diff
@@ -218,4 +258,8 @@ export default {
   validateOneOf,
   validateString,
   validateUint32,
+
+  // Zlib specific
```
Comment (npaun) [inline src/node/internal/validators.ts:262]: I think I'd either remove this comment, or move the functions to the zlib file.

File: src/workerd/api/node/zlib-util.h [hunk 4]
```diff
@@ -1,57 +1,261 @@
...
+#define V(code) if (err == code) return #code;
+  ZLIB_ERROR_CODES(V)
+#undef V
+  return "Z_UNKNOWN_ERROR";
+}
+#endif  // ZLIB_ERROR_CODES
+
+// Certain zlib constants are defined by Node.js itself
+static constexpr auto Z_MIN_CHUNK = 64;
+static constexpr auto Z_MAX_CHUNK = 128 * 1024;
```
Comment (npaun) [inline src/workerd/api/node/zlib-util.h:40]:
```suggestion
static constexpr auto Z_MAX_CHUNK = 128 * 1024 * 1024;
```

This is fixed in my PR too but just pointing it out here

File: src/workerd/api/node/zlib-util.h [hunk 5]
```diff
@@ -1,57 +1,261 @@
...
+  GUNZIP,
+  DEFLATERAW,
+  INFLATERAW,
+  UNZIP,
+  BROTLI_DECODE,
+  BROTLI_ENCODE
+};
+
+struct CompressionError {
+  CompressionError(kj::StringPtr _message, kj::StringPtr _code, int _err)
```
Comment (npaun) [inline src/workerd/api/node/zlib-util.h:71]: I don't personally like variable names starting with `_` but if it's a convention elsewhere in our codebase, then ignore me.

File: src/workerd/api/node/zlib-util.h [hunk 6]
```diff
@@ -1,57 +1,261 @@
...
+    stream.avail_in = inputLength;
+    stream.next_in = input.begin();
+    stream.avail_out = outputLength;
+    stream.next_out = output.begin();
+  };
+  int getFlush() const { return flush; };
+  void setFlush(int value) {
+    flush = value;
+  };
+  void getAfterWriteOffsets(kj::ArrayPtr<kj::byte> writeResult) const {
```
Comment (npaun) [inline src/workerd/api/node/zlib-util.h:100]: Is this function designed this way to conform to a NodeJS interface? I'm not a big fan of out-parameters, nor of storing struct fields as an array.

File: src/workerd/api/node/zlib-util.h [hunk 7]
```diff
@@ -1,57 +1,261 @@
...
+};
+
+class ZlibContext final {
+public:
+  ZlibContext() = default;
+
+  KJ_DISALLOW_COPY(ZlibContext);
+
+  void close();
+  void setBuffers(kj::ArrayPtr<kj::byte> input, uint32_t inputLength, kj::ArrayPtr<kj::byte> output, uint32_t outputLength) {
```
Comment (npaun) [inline src/workerd/api/node/zlib-util.h:90]: What is our convention on when it's ok to implement methods inline? My instinct would be to avoid it, because it slows down compilation time, but this is a project-dependent thing.

File: src/workerd/api/node/zlib-util.c++ [hunk 8]
```diff
@@ -1,11 +1,373 @@
...
 #include "zlib-util.h"
 
 namespace workerd::api::node {
+
 uint32_t ZlibUtil::crc32Sync(kj::Array<kj::byte> data, uint32_t value) {
   // Note: Bytef is defined in zlib.h
   return crc32(value, reinterpret_cast<const Bytef*>(data.begin()), data.size());
 }
+
+void ZlibContext::initialize(int _level, int _windowBits, int _memLevel, int _strategy, jsg::Optional<kj::Array<kj::byte>> _dictionary) {
```
Comment (npaun) [inline src/workerd/api/node/zlib-util.c++:15]: Just commenting to note that this function is the same for both one-shot and async and we should see if we can reuse it.

File: src/workerd/api/node/zlib-util.c++ [hunk 9]
```diff
@@ -1,11 +1,373 @@
...
+          // Make it possible for After() to tell a bad dictionary from bad
+          // input.
+          err = Z_NEED_DICT;
+        }
+      }
+
+      while (stream.avail_in > 0 &&
+             mode == ZlibMode::GUNZIP &&
+             err == Z_STREAM_END &&
+             stream.next_in[0] != 0x00) {
```
Comment (npaun) [inline src/workerd/api/node/zlib-util.c++:243]:
```suggestion
             stream.next_in[0] != '\0') {
```

nit: I think this is more common when we working with bytes

File: src/workerd/api/node/zlib-util.c++ [hunk 10]
```diff
@@ -1,11 +1,373 @@
...
+  // of room.  If there was avail_out left over, then it means
+  // that all the input was consumed.
+  switch (mode) {
+    case ZlibMode::DEFLATE:
+    case ZlibMode::GZIP:
+    case ZlibMode::DEFLATERAW:
+      err = deflate(&stream, flush);
+      break;
+    case ZlibMode::UNZIP:
+      if (stream.avail_in > 0) {
```
Comment (npaun) [inline src/workerd/api/node/zlib-util.c++:178]: This case is very complex. I would consider to extract it into its own function and invoke it from here.

File: src/node/internal/zlib.d.ts [hunk 11]
```diff
@@ -113,3 +115,41 @@ export const CONST_BROTLI_DECODER_ERROR_ALLOC_RING_BUFFER_1: number;
...
+  dictionary?: ArrayBufferView | undefined; // deflate/inflate only, empty dictionary by default
+  info?: boolean | undefined;
+  maxOutputLength?: number | undefined;
+}
+
+type ErrorHandler = (errno: number, code: string, message: string) => void;
+type ProcessHandler = () => void;
+
+export class ZlibStream {
+  public [owner_symbol]: Zlib;
```
[the remark this answers is not in the document]
Comment (anonrig) [reply src/node/internal/zlib.d.ts:136]: Before private properties in javascript (`this.#private_property`) were a thing, people used Symbol's to store private properties. Each symbol is unique and therefore is different from each other.

```
const valSymbol = Symbol('hello');

class Hello {
  setAttr(val) {
    this[valSymbol] = val;
  }

  getAttr() { return this[valSymbol]; }
}

Since zlib was implemented before private properties, this symbol is used to store the variable reference to the parent class instance.

File: src/workerd/api/node/zlib-util.h (hunk 5 above)
Comment (anonrig) [reply src/workerd/api/node/zlib-util.h:71]: This is a pretty common Node.js convention, but I'm pretty sure @jasnell uses it a lot as well. How would you initialize CompressionError where there is a `message` attribute? How would you name this variable int that scenario?

File: src/workerd/api/node/zlib-util.h (hunk 6 above)
Comment (anonrig) [reply src/workerd/api/node/zlib-util.h:100]: I tried to not diverge from Node.js implementation as much as I can to make `maintenance` of this module a lot easier, unless there is a specific reason to do so.

File: src/workerd/api/node/zlib-util.h (hunk 7 above)
Comment (anonrig) [reply src/workerd/api/node/zlib-util.h:90]: i'm not sure. i have this way in my mind that if the function is simple to implement in header, i put it in header. but technically we should have `zlib-util-inl.h` class if we want to conform to Google code style. @jasnell?

File: src/workerd/api/node/zlib-util.h (hunk 5 above)
Comment (npaun) [reply src/workerd/api/node/zlib-util.h:71]: Same name for both is what I'm used to, but if we're doing underscores I'll change.

File: src/node/.eslintrc.json (hunk 1 above)
Comment (anonrig) [reply src/node/.eslintrc.json:15]: Agree: https://github.com/cloudflare/workerd/pull/2554

File: src/node/internal/internal_zlib_base.ts [hunk 12]
```diff
@@ -0,0 +1,558 @@
...
+  let offset = self._outOffset;
+  const chunkSize = self._chunkSize;
+
+  let error;
+  self.on('error', function onError(er) {
+    error = er;
+  });
+
+  while (true) {
+    // TODO: This was `writeSync` before, but it's not anymore.
```
Comment (jasnell) [inline src/node/internal/internal_zlib_base.ts:211]: Is this our todo or an existing todo from the node.js source?

File: src/workerd/api/node/zlib-util.h [hunk 13]
```diff
@@ -1,57 +1,204 @@
 // Copyright (c) 2017-2022 Cloudflare, Inc.
 // Licensed under the Apache 2.0 license found in the LICENSE file or at:
 //     https://opensource.org/licenses/Apache-2.0
-
+// Copyright Joyent and Node contributors. All rights reserved. MIT license.
 #include <workerd/jsg/jsg.h>
```
Comment (jasnell) [inline src/workerd/api/node/zlib-util.h:5]:
```suggestion
#pragma once
#include <workerd/jsg/jsg.h>
```

File: src/workerd/api/node/zlib-util.h (hunk 5 above)
Comment (jasnell) [reply src/workerd/api/node/zlib-util.h:71]: For this code that is largely adopted from Node.js I prefer that we maintain the Node.js conventions in order to make porting (and more importantly updating later) easier

File: src/workerd/api/node/zlib-util.h (hunk 7 above)
Comment (jasnell) [reply src/workerd/api/node/zlib-util.h]: There may be some difference of opinion on it in the team but for me, it really comes down to the complexity of the function. Generally try to avoid inlining in the headers unless the body is very trivial (e.g. a trivial constructor, or a trivial getter/setter). When in doubt, if it's going to be more than a few lines, don't inline.

**Comment by jasnell** [comment]
Overall looking good so far. Definitely want to see more tests land with this tho so will hold off signing off until we get those.

File: src/node/internal/internal_zlib_base.ts (hunk 12 above)
Comment (anonrig) [reply src/node/internal/internal_zlib_base.ts:211]: This is our TODO. I implemented the `sync` path, not didn't implement the async path on C++.

File: src/workerd/api/node/zlib-util.h (hunk 13 above)
Comment (anonrig) [reply src/workerd/api/node/zlib-util.h:5]: Header files should only be included only once. Why do we need to have a pragma once here? Do we not trust the compiler to include a header file only once, or is it something else?

Comment (npaun) [reply src/workerd/api/node/zlib-util.h:5]: The default behaviour is to include every time. `#pragma once` tells the compiler not to include `zlib-util.h` more than once.

File: src/workerd/api/node/zlib-util.h (hunk 7 above)
Comment (npaun) [reply src/workerd/api/node/zlib-util.h]: @jasnell That's my usual approach as well. However wouldn't we get faster builds if we totally crack down on inline implementations?

File: src/node/internal/internal_zlib_base.ts (hunk 12 above)
Comment (jasnell) [reply src/node/internal/internal_zlib_base.ts:211]: Ok, to differentiate, make the todo like... This will help since the `(soon)` is our convention vs. node.js convention of `TODO(username): `

```suggestion
    // TODO(soon): This was `writeSync` before, but it's not anymore.
```

File: src/workerd/api/node/zlib-util.h [hunk 14]
```diff
@@ -1,57 +1,206 @@
...
+  z_stream stream{};
+};
+
+using CompressionStreamErrorHandler = jsg::Function<void(int, kj::StringPtr, kj::StringPtr)>;
+
+template <typename CompressionContext>
+class CompressionStream {
+public:
+  CompressionStream() = default;
+  ~CompressionStream();
```
Comment (jasnell) [inline src/workerd/api/node/zlib-util.h:146]:
```suggestion
  ~CompressionStream() noexcept(false);
```

File: src/workerd/api/node/zlib-util.h [hunk 15]
```diff
@@ -1,57 +1,206 @@
...
-    BROTLI_ENCODE
+  class ZlibStream final : public jsg::Object, public CompressionStream<ZlibContext> {
+  public:
+    ZlibStream(ZlibMode mode);
+    KJ_DISALLOW_COPY_AND_MOVE(ZlibStream);
+    static jsg::Ref<ZlibStream> constructor(ZlibModeValue mode);
+
+    // Instance methods
+    void initialize(int windowBits, int level, int memLevel, int strategy, kj::Array<kj::byte> writeState, jsg::Function<void()> writeCallback, jsg::Optional<kj::Array<kj::byte>> dictionary);
+    void write(jsg::Lock& js,int flush, kj::Array<kj::byte> input, int inputOffset, int inputLength, kj::Array<kj::byte> output, int outputOffset, int outputLength);
```
Comment (jasnell) [inline src/workerd/api/node/zlib-util.h:185]:
```suggestion
    void write(jsg::Lock& js, int flush, kj::Array<kj::byte> input, int inputOffset, int inputLength, kj::Array<kj::byte> output, int outputOffset, int outputLength);
```

File: src/node/zlib.ts [hunk 16]
```diff
@@ -1,8 +1,74 @@
+import * as zlib from 'node-internal:internal_zlib';
 import { crc32, constants } from 'node-internal:internal_zlib';
+import { nodeJsZlib } from 'workerd:compatibility-flags';
 
-export { crc32, constants };
+function protectMethod(method: unknown): unknown {
+  if (nodeJsZlib) {
+    return function notImplemented() {
+      throw new Error('Compatibility flag "nodejs_compat_zlib" is not enabled');
```
Comment (jasnell) [inline src/node/zlib.ts:8]:
```suggestion
      throw new Error('Compatibility flag "nodejs_zlib" is not enabled');
```

File: src/workerd/io/compatibility-date.capnp [hunk 17]
```diff
@@ -580,4 +580,10 @@ struct CompatibilityFlags @0x8f8c1b68151b6cef {
           packages = "2024-03-01", backport = 0)
     $experimental;
   # Enables Python Workers and uses the bundle from the Pyodide source directory directly. For testing only.
+
+  nodeJsZlib @59 :Bool
+      $compatEnableFlag("nodejs_zlib")
+      $compatDisableFlag("no_nodejs_zlib")
+      $experimental;
+  # Enables node:zlib implementation while it is in-development.
```
Comment (jasnell) [inline src/workerd/io/compatibility-date.capnp:588]:
```suggestion
  # Enables node:zlib implementation while it is in-development.
  # Once the node:zlib implementation is complete, this will be automatically enabled when
  # nodejs_compat is enabled.
```

**Review by jasnell — APPROVED** [review]
Looks good once we're sure the internal CI is green.

File: src/workerd/api/node/zlib-util.h [hunk 18]
```diff
@@ -1,57 +1,245 @@
...
+  bool initializeZlib();
+  kj::Maybe<CompressionError> setDictionary();
+
+  CompressionError constructError(kj::StringPtr message) const {
+    if (stream.msg != nullptr) message = kj::StringPtr(stream.msg);
+
+    return {kj::str(message), kj::str(ZlibStrerror(err)), err};
+  };
+
+  kj::MutexGuarded<bool> initialized{false};
```
Comment (mikea) [inline src/workerd/api/node/zlib-util.h:145]: this is rather unusual: please leave a comment which threads access it and why this synchronization is necessary.

**Review by mikea — APPROVED** [review]
appreciate the effort to make all .ts code typed.

File: src/workerd/api/node/zlib-util.h (hunk 18 above)
Comment (jasnell) [reply src/workerd/api/node/zlib-util.h:145]: @anonrig and I discussed this a bit yesterday. I don't think this is actually needed and it should be removed in the next iteration as this moves forward. A mutex is used here in Node.js because there are cases that need to be protected from multiple threads accessing this but for workerd it shouldn't be necessary.

**Review by npaun — CHANGES_REQUESTED** [review]
Tests are failing under ASan. I'm worried there might be some kind of corruption.

File: src/workerd/api/node/zlib-util.h (hunk 18 above)
Comment (anonrig) [reply src/workerd/api/node/zlib-util.h]: Yeah, it is unnecessary. While investigating the ASAN build, I removed it.

**Comment by npaun** [comment]
Do we know why the unnecessary mutex guard caused the test failure?

It doesn't immediately jump out at me what this has to do with a mutex
`Fatal uncaught kj::Exception: workerd/api/node/zlib-util.c++:311: failed: expected status == 0 || status == (-3); jsg.Error: Uncaught error on closing zlib stream`

**Comment by anonrig** [comment]
> Do we know why the unnecessary mutex guard caused the test failure?

Yes, the class destructor was checking the reference not the value, and causing the sanitizer to fail.

**Review by npaun — APPROVED** [review]

**Comment by npaun** [comment]
Aside: That implicit conversion to `operator T*` in libkj is kind of wack.
